File: src/37_SerializeBinaryTrees/37_SerializeBinaryTrees.cpp
// 面试题37：序列化二叉树
// 题目：请实现两个函数，分别用来序列化和反序列化二叉树。

// 输入输出格式与 LeetCode 目前使用的方式一致，例子：
// 输入：root = [1,2,3,null,null,4,5]
// 输出：[1,2,3,null,null,4,5]

// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));

#include <iostream>
#include <queue>
#include <string>
#include <vector>

#include "../Utilities/BinaryTree.h"

using namespace std;

using TreeNode = BinaryTreeNode;

class Codec {
   public:
    // Encodes a tree to a single string.
    string serialize( TreeNode* root ) {
        if ( !root ) return "[]";
        string result = "[";

        // 层序遍历
        queue<TreeNode*> q;
        q.push( root );
        while ( !q.empty() ) {
            TreeNode* node = q.front();
            q.pop();
            // 如果节点非空
            if ( node ) {
                // 使用std::to_string函数将数值转换为字符串，并在其后添上逗号
                result += ( std::to_string( node->val ) + "," );
                // 将左右子节点入列，无论是否为空
                q.push( node->left );
                q.push( node->right );
            } else {
                result += "null,";
            }
        }
        result.pop_back();  // 去掉末尾的逗号
        result += "]";
        return result;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize( string data ) {
        if ( data == "[]" ) return nullptr;

        // 去掉首尾括号后，将原字符串拆分为列表
        vector<string> vals = split( data.substr( 1, data.length() - 2 ) );
        // 生成新的根节点，利用std::stoi将字符串转换为整数，
        // 此时根节点不为空，因此可以安全转换
        TreeNode* root = new TreeNode( std::stoi( vals[ 0 ] ) );

        queue<TreeNode*> q;
        q.push( root );
        int i = 1;  // 用以指示下一节点
        while ( !q.empty() ) {
            TreeNode* node = q.front();
            q.pop();
            // 左子节点
            if ( vals[ i ] != "null" ) {
                node->left = new TreeNode( std::stoi( vals[ i ] ) );
                q.push( node->left );
            }
            i++;
            // 右子节点
            if ( vals[ i ] != "null" ) {
                node->right = new TreeNode( std::stoi( vals[ i ] ) );
                q.push( node->right );
            }
            i++;
        }
        return root;
    }

   private:
    // 将字符串拆分为列表
    vector<string> split( string s, char sep = ',' ) {
        vector<string> strList;
        int i = 0, j = 0, length = s.length();
        while ( i < length ) {
            while ( j < length && s[ j ] != sep ) j++;
            strList.push_back( s.substr( i, j - i ) );
            i = ++j;
        }
        return strList;
    }
};

class Test {
   public:
    void checkResult( const string testName, TreeNode* root, string data ) {
        cout << testName << ": ";

        string resultStr = solution.serialize( root );
        TreeNode* resultRoot = solution.deserialize( resultStr );

        bool sameTree = isSameTree( root, resultRoot );

        cout << "sameStr=" << ( ( data == resultStr ) ? "true, " : "false, " )
             << "sameTree=" << ( sameTree ? "true." : "false." );
        cout << endl;
    }

    //            8
    //        6      10
    //       5 7    9  11
    void Test1() {
        string data =
            "[8,6,10,5,7,9,11,null,null,null,null,null,null,null,null]";

        ValVec vals{ 8, 6, 10, 5, 7, 9, 11 };
        NodeVec nodes = createNodesVec( vals );

        connectNodesIndex( nodes, 0, 1, 2 );
        connectNodesIndex( nodes, 1, 3, 4 );
        connectNodesIndex( nodes, 2, 5, 6 );

        checkResult( "Test1", nodes[ 0 ], data );
        destroyTree( nodes[ 0 ] );
    }

    //            5
    //          4
    //        3
    //      2
    void Test2() {
        string data = "[5,4,null,3,null,2,null,null,null]";

        ValVec vals{ 5, 4, 3, 2 };
        NodeVec nodes = createNodesVec( vals );

        connectNodesIndex( nodes, 0, 1, -1 );
        connectNodesIndex( nodes, 1, 2, -1 );
        connectNodesIndex( nodes, 2, 3, -1 );

        checkResult( "Test2", nodes[ 0 ], data );
        destroyTree( nodes[ 0 ] );
    }

    //        5
    //         4
    //          3
    //           2
    void Test3() {
        string data = "[5,null,4,null,3,null,2,null,null]";

        ValVec vals{ 5, 4, 3, 2 };
        NodeVec nodes = createNodesVec( vals );

        connectNodesIndex( nodes, 0, -1, 1 );
        connectNodesIndex( nodes, 1, -1, 2 );
        connectNodesIndex( nodes, 2, -1, 3 );

        checkResult( "Test3", nodes[ 0 ], data );
        destroyTree( nodes[ 0 ] );
    }

    //        5
    //         5
    //          5
    //         5
    //        5
    //       5 5
    //      5   5
    void Test4() {
        string data =
            "[5,null,5,null,5,5,null,5,null,5,5,5,null,null,5,null,null,null,"
            "null]";

        ValVec vals{ 5, 5, 5, 5, 5, 5, 5, 5, 5 };
        NodeVec nodes = createNodesVec( vals );

        connectNodesIndex( nodes, 0, -1, 1 );
        connectNodesIndex( nodes, 1, -1, 2 );
        connectNodesIndex( nodes, 2, 3, -1 );
        connectNodesIndex( nodes, 2, 3, -1 );
        connectNodesIndex( nodes, 3, 4, -1 );
        connectNodesIndex( nodes, 4, 5, 6 );
        connectNodesIndex( nodes, 5, 7, -1 );
        connectNodesIndex( nodes, 6, -1, 8 );

        checkResult( "Test4", nodes[ 0 ], data );
        destroyTree( nodes[ 0 ] );
    }

    // 只有一个节点
    void Test5() {
        string data = "[1,null,null]";

        ValVec vals{ 1 };
        NodeVec nodes = createNodesVec( vals );

        checkResult( "Test5", nodes[ 0 ], data );
        destroyTree( nodes[ 0 ] );
    }

    // 没有节点
    void Test6() { checkResult( "Test6", nullptr, "[]" ); }

    void runTest() {
        Test1();
        Test2();
        Test3();
        Test4();
        Test5();
        Test6();
    }

   private:
    Codec solution;

    bool isSameTree( TreeNode* root1, TreeNode* root2 ) {
        if ( !root1 && !root2 ) return true;
        if ( !root1 || !root2 || root1->val != root2->val ) return false;
        return isSameTree( root1->left, root2->left ) &&
               isSameTree( root1->right, root2->right );
    }
};

int main() {
    Test test;
    test.runTest();
    return 0;
}

File: src/37_SerializeBinaryTrees/37_SerializeBinaryTrees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

import os
import sys

sys.path.append(os.path.split(sys.path[0])[0])


from Utilities import binary_tree
from Utilities.binary_tree import TreeNode


class Codec:
    def serialize(self, root: TreeNode) -> str:
        """Encodes a tree to a single string.

        :type root: TreeNode
        :rtype: str
        """
        if not root:
            return "[]"

        result, q = [], [root]
        while q:
            node = q.pop(0)
            if node:
                result.append(str(node.val))
                q.append(node.left)
                q.append(node.right)
            else:
                result.append("null")
            # Python的join函数，可将列表直接转换成字符串
        return "[" + ",".join(result) + "]"

    def deserialize(self, data: str) -> TreeNode:
        """Decodes your encoded data to tree.

        :type data: str
        :rtype: TreeNode
        """
        if data == "[]":
            return None

        vals = data[1:-1].split(",")
        root = TreeNode(int(vals[0]))

        q, i = [root], 1
        while q:
            node = q.pop(0)
            # 左子节点
            if vals[i] != "null":
                node.left = TreeNode(int(vals[i]))
                q.append(node.left)
            i += 1
            # 右子节点
            if vals[i] != "null":
                node.right = TreeNode(int(vals[i]))
                q.append(node.right)
            i += 1
        return root


class Test:
    def __init__(self) -> None:
        self.solution = Codec()

    def checkResult(self, testName: str, root: TreeNode, data: str):
        text = f"{testName}: "

        resultStr = self.solution.serialize(root)
        resultRoot = self.solution.deserialize(resultStr)

        text += f"sameStr={data==resultStr}, "
        text += f"sameTree={self.isSameTree(root, resultRoot)}."
        print(text)

    def isSameTree(self, root1: TreeNode, root2: TreeNode):
        if not root1 and not root2:
            return True
        if not root1 or not root2 or root1.val != root2.val:
            return False
        return self.isSameTree(root1.left, root2.left) and self.isSameTree(
            root1.right, root2.right
        )

    #            8
    #        6      10
    #       5 7    9  11
    def Test1(self):
        data = "[8,6,10,5,7,9,11,null,null,null,null,null,null,null,null]"

        vals = [8, 6, 10, 5, 7, 9, 11]
        nodes = binary_tree.createTreeNodesVec(vals)

        idxs = [
            (0, 1, 2),
            (1, 3, 4),
            (2, 5, 6),
        ]
        binary_tree.connectTreeNodesIndexVec(nodes, idxs)

        self.checkResult("Test1", nodes[0], data)

    #            5
    #          4
    #        3
    #      2
    def Test2(self):
        data = "[5,4,null,3,null,2,null,null,null]"

        vals = [5, 4, 3, 2]
        nodes = binary_tree.createTreeNodesVec(vals)

        idxs = [
            (0, 1, -1),
            (1, 2, -1),
            (2, 3, -1),
        ]
        binary_tree.connectTreeNodesIndexVec(nodes, idxs)

        self.checkResult("Test2", nodes[0], data)

    #        5
    #         4
    #          3
    #           2
    def Test3(self):
        data = "[5,null,4,null,3,null,2,null,null]"

        vals = [5, 4, 3, 2]
        nodes = binary_tree.createTreeNodesVec(vals)

        idxs = [
            (0, -1, 1),
            (1, -1, 2),
            (2, -1, 3),
        ]
        binary_tree.connectTreeNodesIndexVec(nodes, idxs)

        self.checkResult("Test3", nodes[0], data)

    #        5
    #         5
    #          5
    #         5
    #        5
    #       5 5
    #      5   5
    def Test4(self):
        data = "[5,null,5,null,5,5,null,5,null,5,5,5,null,null,5,null,null,null,null]"

        vals = [5, 5, 5, 5, 5, 5, 5, 5, 5]
        nodes = binary_tree.createTreeNodesVec(vals)

        idxs = [
            (0, -1, 1),
            (1, -1, 2),
            (2, 3, -1),
            (2, 3, -1),
            (3, 4, -1),
            (4, 5, 6),
            (5, 7, -1),
            (6, -1, 8),
        ]
        binary_tree.connectTreeNodesIndexVec(nodes, idxs)

        self.checkResult("Test4", nodes[0], data)

    # 只有一个节点
    def Test5(self):
        data = "[1,null,null]"

        vals = [1]
        nodes = binary_tree.createTreeNodesVec(vals)

        self.checkResult("Test5", nodes[0], data)

    # 没有节点
    def Test6(self):
        self.checkResult("Test6", None, "[]")

    def runTest(self):
        self.Test1()
        self.Test2()
        self.Test3()
        self.Test4()
        self.Test5()
        self.Test6()


if __name__ == "__main__":
    test = Test()
    test.runTest()
